adapter/internal/sourcewatcher/sourcewatcher.go

```diff
@@ -0,0 +1,365 @@
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error while authenticating with the remote repository : %s", err.Error()),
+			Severity: logging.CRITICAL,
+			ErrorCode: 2502,
+		})
+	}
+
+	pullOptions := &git.PullOptions{
+		Auth: gitAuth,
+	}
+
+	if branch != "" {
+		pullOptions.ReferenceName = plumbing.ReferenceName("refs/heads/" + branch)
+	}
+
+	workTree, err := localRepository.Worktree()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error while retrieving local repository worktree : %s", err.Error()),
+			Severity: logging.MAJOR,
+			ErrorCode: 2508,
+		})
+	}
+
+	err = workTree.Pull(pullOptions)
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error while pulling changes from the remote repository : %s", err.Error()),
+			Severity: logging.MAJOR,
+			ErrorCode: 2509,
+		})
+	}
+}
+
+// processArtifactChanges undeploy the APIs whose artifacts are not present in the repository
+func processArtifactChanges() (err error){
+	conf, err := config.ReadConfigs()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error reading configs : %s", err.Error()),
+			Severity: logging.BLOCKER,
+			ErrorCode: 2500,
+		})
+		return err
+	}
+
+	apisDirName := filepath.FromSlash(conf.Adapter.SourceControl.ArtifactsDirectory + "/" + apisArtifactDir)
+	files, err := ioutil.ReadDir(apisDirName)
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error while reading API artifacts directory : %s", err.Error()),
+			Severity: logging.MAJOR,
+			ErrorCode: 2510,
+		})
+	}
+
+	// Assign the API artifacts to a temporary map
+	removedArtifacts := artifactsMap
+
+	for _, apiProjectFile := range files {
+		if apiProjectFile.IsDir() {
+			// If the artifact is present in the artifacts directory, remove it from the temporary map
+			if _, ok := removedArtifacts[apiProjectFile.Name()]; ok {
+				delete(removedArtifacts, apiProjectFile.Name())
+			}
+			continue
+		} else if !strings.HasSuffix(apiProjectFile.Name(), zipExt) {
+			continue
+		}
+		// If the zip artifact is present in the artifacts directory, remove it from the temporary map
+		if _, ok := removedArtifacts[apiProjectFile.Name()]; ok {
+			delete(removedArtifacts, apiProjectFile.Name())
+		}
+	}
+
+	// Undeploy the APIs whose artifacts are not present in the repository
+	for _, apiProject := range removedArtifacts {
+		apiYaml := apiProject.APIYaml.Data
+
+		vhostToEnvsMap := make(map[string][]string)
+		for _, environment := range apiProject.Deployments {
+			vhostToEnvsMap[environment.DeploymentVhost] =
+				append(vhostToEnvsMap[environment.DeploymentVhost], environment.DeploymentEnvironment)
+		}
+
+		for vhost, environments := range vhostToEnvsMap {
+			if vhost == "" {
+				// ignore if vhost is empty, since it deletes all vhosts of API
```

Comment:
Can you please elaborate more on this? can vhost be empty?

Reply:
Fixed with 5eadc80. It was used earlier when using the `DeleteAPIsWithUUID` method for testing purposes. Now the `DeleteAPIs` method is being used.